distributed/protocol.py

```diff
@@ -64,35 +62,77 @@
     default_compression = 'lz4'
 
 
+def _extract_big_bytes(x, big, path=()):
+    if type(x) is dict:
+        for k, v in x.items():
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
+                del x[k]
+    elif type(x) is list:
+        for k, v in enumerate(x):
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
+                x[k] = None
+
+
 def dumps(msg):
```

Comment:
Timings from the array benchmark (normalized for same number of calls, since they slightly differ between runs).

```
This PR:
- dumps cumulative time: 0.6042 s
- loads cumulative time: 0.5028 s

Master:
- dumps cumulative time: 0.5738 s
- loads cumulative time: 0.3064 s
```